pkg/daemon/update.go

```diff
@@ -83,6 +88,146 @@ func reloadService(name string) error {
+	}
+
+	var inDesiredConfig bool
+	var missingODC bool
+	if missingODC, inDesiredConfig, err = dn.updateConfigAndState(state); err != nil {
+		return fmt.Errorf("could not apply update: setting node's state to Done failed. Error: %w", err)
+	}
+
+	if missingODC {
+		return fmt.Errorf("error updating state.currentconfig from on-disk currentconfig")
+	}
+
+	if inDesiredConfig {
+		// (re)start the config drift monitor since rebooting isn't needed.
+		dn.startConfigDriftMonitor()
+		return nil
+	}
+
+	// currentConfig != desiredConfig, kick off an update
+	return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig, true)
+}
+
+func (dn *Daemon) executeReloadServiceNodeDisruptionAction(serviceName string, reloadErr error) error {
+	if reloadErr != nil {
+		if dn.nodeWriter != nil {
+			dn.nodeWriter.Eventf(corev1.EventTypeWarning, "FailedServiceReload", fmt.Sprintf("Reloading service %s failed. Error: %v", serviceName, reloadErr))
+		}
+		return fmt.Errorf("could not apply update: reloading %s configuration failed. Error: %w", serviceName, reloadErr)
+	}
+
+	err := upgrademonitor.GenerateAndApplyMachineConfigNodes(
+		&upgrademonitor.Condition{State: mcfgalphav1.MachineConfigNodeUpdatePostActionComplete, Reason: string(mcfgalphav1.MachineConfigNodeUpdateReloaded), Message: fmt.Sprintf("Node has reloaded service %s", serviceName)},
+		&upgrademonitor.Condition{State: mcfgalphav1.MachineConfigNodeUpdateReloaded, Reason: fmt.Sprintf("%s%s", string(mcfgalphav1.MachineConfigNodeUpdatePostActionComplete), string(mcfgalphav1.MachineConfigNodeUpdateReloaded)), Message: fmt.Sprintf("Upgrade required a service %s reload. Completed this this as a post update action.", serviceName)},
+		metav1.ConditionTrue,
+		metav1.ConditionTrue,
+		dn.node,
+		dn.mcfgClient,
+		dn.featureGatesAccessor,
+	)
+	if err != nil {
+		klog.Errorf("Error making MCN for Reloading success: %v", err)
+	}
+
+	if dn.nodeWriter != nil {
+		dn.nodeWriter.Eventf(corev1.EventTypeNormal, "ServiceReload", "Config changes do not require reboot. Service %s was reloaded.", serviceName)
+	}
+	logSystem("%s service reloaded successfully!", serviceName)
+	return nil
+}
+
+// performPostConfigChangeNodeDisruptionAction takes action based on the cluster's Node disruption policies.
+// For non-reboot action, it applies configuration, updates node's config and state.
+// In the end uncordon node to schedule workload.
+// If at any point an error occurs, we reboot the node so that node has correct configuration.
+func (dn *Daemon) performPostConfigChangeNodeDisruptionAction(postConfigChangeActions []opv1.NodeDisruptionPolicyStatusAction, configName string) error {
+
+	logSystem("Executing performPostConfigChangeNodeDisruptionAction(drain already complete/skipped) for config %s", configName)
+	for _, action := range postConfigChangeActions {
+		logSystem("Executing postconfig action: %v for config %s", action.Type, configName)
+		if action.Type == opv1.RebootStatusAction {
```

Comment:
I wonder if we should make reboot/none a separate condition, i.e. if RebootStatusActions is in postConfigChangeActions we skip all other actions, and none can only exist if others do not?

Actually reading the rest of the PR, it looks like you have it set up such that if RebootStatusActions exists, is the list always just that one object so we are implicitly covered? (i.e. you can't have postConfigChangeActions contain both reboot and other actions like reload, and none shows up only if no other actions exist?)

Reply:
Yeah the list of actions is built like this:
- if there is atleast one Reboot action, return a list of just a singular Reboot Action
- if there is a None action and other actions, strip out any None actions
- if there is only None actions in the list, return a list of just a singular None Action

Sorry if I made that a little convoluted, I tried to make the comments make sense!